Dev light improvements

ron-adam opened the pull request:
Proposals for improving lights and materials.

This pull request is a preliminary work in progress to improve the lights and materials in Datoviz. 
(It is not  a complete proposal but is far enough along to allow for testing and discussion for further development.)

These items are interrelated so have been grouped together as they are related.

The following files contain the main changes.  Alterations in other files are so that these changes can be used and tested.
(lighting.glsl,  graphics_mesh.vert, graphics_mesh.frag, graphics_sphere.vert, and graphics_sphere.frag)

1.  Separate the lighting code into separate glsl files to be included in shaders.
     Questions:   Should the lighting.glsl be split into two files.  One for each lighting mode.  Should the binding definitions be included with it or be elsewhere?

2.  Fixed materials so they are not tied to separate lights.  Materials should react the same to any lights.

3.  Use normalized parameter values for materials parameters.  (There are examples using normalized values.)

4.  Propose moving set_light_color and set_light_dir functions to panels.  (Not completed yet.)
    
5.  Add either a set_light_pos function, or a flag to select which behavior on set_light_dir.   (Not complete yet.)

6.  Created an mesh light example file to explore how these changes work.

7.  Modified python mesh and sphere examples so these changes can be viewed.

Still to do:

   -  C example files still need updating.  (along with visual and panel C files.)
   -  Binding definitions need to be put in place.
   -  Shaders will need to be changed from current work arounds used for testing to use the new bindings.
   -  Python files will need to be updated to use the newer API.
   
The purpose of this pull request is to share these possible upgrades for further discussion and deciding how the details can be implemented.  (or if some should not be.)

Notes:
     The current behavior for testing uses light positions instead of light directions.  This is done in the fragment shader, and is only meant to be a temporary work around. 
     
Please feedback on even small details will greatly be appreciated.

rossant commented:
Excellent work!

As a general comment, I’d say we have several options moving forward, considering the current timeline:

1. Do things as properly as possible: this would require waiting for v0.3 to be released first, then starting work on v0.4. That version would involve revamping the internal rendering protocol (`request.h`) to support additional Vulkan features needed to properly handle multiple lighting models.

2. Try to fit this proposal into the current architecture before releasing v0.3, which would require workarounds and result in a suboptimal implementation that would need to be redone for v0.4.

3. Find a middle ground: stick to a single lighting model for now (e.g., the advanced one), but keep the basic lighting code in the shader for now, even if unused. We don’t expose lighting in the panel yet. The only change we might want to make is to define two uniform bindings per visual (mesh and sphere): one for the advanced light and one for the material. We can use the DvzParams API for both. This is suboptimal because if multiple mesh/sphere visuals are added to the same panel, the same lighting values will have to be duplicated. If they don’t match, the result will be very unexpected. However, this would allow us to still merge this PR before releasing v0.3 and have support for multiple lights and different materials.

I’d lean toward option (3). What do you think?

Now to answer your specific questions:

Q1. Let’s have a single `lighting.glsl` file. We won’t put the binding definitions there, but we can define the `BasicLight` and `AdvancedLight` structs in it, like:

```glsl
struct BasicLight {
    vec4 pos;
    vec4 color;
    vec4 params;
};
````

Then in fragment shaders, we can do:

```glsl
layout(std140, binding = 0) uniform light {
    BasicLight basic;
};
```

And in the shader code:

```glsl
light.basic.pos
```

Q4. We can leave this for after v0.3.

Q5. As a temporary solution, we could have `set_light_pos()` accept a keyword argument like `is_dir=True`, and the Python binding could set the alpha component accordingly.

That way, we only introduce minimal changes to the user-facing API for now. From your PR, it looks like you’ve implemented it without touching the C code or Python wrapper, which is great!

If we go with option (3), the main remaining change before merging and releasing v0.3 would be to update `mesh.c` and `sphere.c` to add two new uniform bindings to the mesh and sphere visuals, and update the `light_pos`/`light_params` C functions and their Python counterparts using separate `DvzParams` instances per binding.

For example, for the mesh, we could define the following slots:

```c
// Common slots (hidden in `_common_setup()`)
dvz_visual_slot(visual, 0, DVZ_SLOT_DAT); // MVP
dvz_visual_slot(visual, 1, DVZ_SLOT_DAT); // viewport

// Mesh-specific slots
dvz_visual_slot(visual, 2, DVZ_SLOT_DAT); // light pos, dir, color
dvz_visual_slot(visual, 3, DVZ_SLOT_DAT); // material params
dvz_visual_slot(visual, 4, DVZ_SLOT_DAT); // contour params
dvz_visual_slot(visual, 5, DVZ_SLOT_TEX); // texture
```

Then bind with something like:

```c
dvz_visual_params(visual, 2, sizeof(DvzAdvancedLight));
dvz_visual_params(visual, 3, sizeof(DvzMaterial));
// etc
```

The binding indices should probably be defined as `#define` at the top of the C file, like `#define DVZ_MESH_SLOT_LIGHT 2` etc.

The GLSL structures defined in `lighting.glsl` will have to be duplicated exactly in the C code.

Also, as you noticed, we should rename `set_light_params()` to `set_material_params()`.

Similarly with the sphere visual.

Would you prefer to implement these changes and finish the PR, or would you like me to handle it? I still have a few days of work left on the documentation. Ideally, I’d like to release v0.3 by May 26 at the latest.

rossant commented on examples/features/mesh_light.py:
```diff
@@ -0,0 +1,83 @@
+    has_changed |= dvz.gui_slider_vec3('Pos 2 XYZ', -10, +10, light_pos[2])
+    has_changed |= dvz.gui_slider_vec3('Pos 3 XYZ', -10, +10, light_pos[3])
+
+    has_changed |= dvz.gui_slider_vec3('Color 0 RGB', 0, 1, light_color[0])
```
You could try to separate the sets of sliders with something like (untested):

```python
dvz.gui_text("Light positions:")
```

rossant commented on src/scene/glsl/graphics_mesh.frag:
```diff
@@ -141,80 +142,45 @@ void main()
-        for (int i = 0; i < 4; i++)
+        // TODO:  Update visual and panel setter functions.
+        // Current visual setters don't set w and a.
+        for (int i=0; i<4; i++)
```
That will be fixed directly in the C functions that set the light and material parameters, they'll use the `dvz_params*()` functions with the appropriate values.

rossant commented on src/scene/glsl/graphics_sphere.frag:
```diff
@@ -66,15 +40,24 @@ void main()
     // Get lighting.  (Reqires LightParams structure to be set.
-    out_color = lighting(pos, in_color, normal, in_cam_pos,
-                         params.light_pos, light_color, params.light_params);
+    out_color = basic_lighting(pos, in_color, material, normal, in_cam_pos, light_pos, light_color);
```
If we go with option (3), we'll likely use advanced lighting here, as in mesh.

ron-adam commented:
I didn't touch the C code due to me not being familiar with the interdependencies yet.  And I figured you could probably make the changes there much faster.  (But with option 3,  it may not require difficult changes.)

Option 3 gets things working so that we can move forward on other areas.   The more in depth changes of the internal rendering protocol may benefit from having a more complete user API.  (Just a hunch as I have not looked at that code yet.)

On Q4,  yes, moving the lights to panels later is a good choice.   Also we may have other options involving how light positions are defined when we add more camera options.  

I was looking at other sources and they use a single float to define shininess,  and where I have the exponent... They have emission.   That is a better choice.  An object is never shiny for red, but dull for blue.  And emission color does not need to match the ambient color or base color.   An added float can set the level of emission, and then only that need to be changed to toggle something from on or off.  example... a rendered light bulb located at a light position. 

I think I can make some additional edits based on your review.   If I have time today I may look at doing the C code changes too.  (or part of it. )

Thanks for the detailed review. :-)

rossant commented:
Got it. Happy to work on the C code changes myself if you prefer, your call!

ron-adam commented:
I tried to make some of the C changes yesterday.  There are more files involved than I was expecting.  This is giving me a better idea of how the C code is organized, so I'll try again today, but divide into smaller changes this today.

ron-adam commented:
I think I'll let you do this part, and I'll study what you do.

I attempted to add a light color field to the sphere object as a test change.  The value set in visuals.py  isn't getting read and changing the new light_color field I added in sphere.c which then should have updated the shader layout/structure.  It compiled and ran, but apparently I was missed a needed changed somewhere. I can see the color change in visuals.py, and set the color manually in sphere.c.   It appears the dvz_sphere_light_color function I added isn't being called when the color is changed in visuals.py.  It is being called once when the program starts.

How are the _ctypes.py files rebuilt?   And the symbol table?

rossant commented:
`just headers` parses the C headers and generates `tools/headers.json` with all the C function/structs etc.

`just ctypes` reads this JSON file and generates `_ctypes.py`

I prefer to implement changes like this in this order:

1. Updating the shaders, C structures, C functions
2. Writing some tests/examples in C
3. Rebuilding the ctypes wrapper
4. Writing Python examples
5. Updating the documentation

Let me know once you're ready for me to take over, and if you can, please write a brief summary of the changes you'd like me to implement!

ron-adam commented:
Yes, I think I had most of that correct.  Although not in that order.

I'll need to go back to step 2.  But I believe I got most of it now.

Just figured this out a few minutes ago....
What's puzzling is in visuals.py,  Some methods for setting values set the value attribute, while others call the dvz function.
To get my new added value to update the structure, I had to do both, which isn't consistent with how it's done in other classes. 

In the mesh class of visuals.py:
```python
    def set_shine(self, value: float) -> None:
        """
        Set the surface shininess.

        Parameters
        ----------
        value : float
            The shine value.
        """
        self.shine = value
        dvz.mesh_shine(self.c_visual, value)
```
Where dvz.mesh_shine() is an added function in mesh.c.

This works, just maybe not in the intended way.   I believe I've got most everything else figured out.
Am I missing a step?

Thanks for the help!

rossant commented:
I'm not really happy with the current implementation of `visuals.py` — there's too much hidden magic. Most setter functions just do `self.xxx = yyy`, which is intercepted by `__setattr__()` and automatically calls something like `dvz.thevisualname_xxx(..., yyy)`. See the `Prop` implementation for reference.

The issue is that this doesn't work in all cases, since some functions require more complex logic. But in your case, it ought to work.

This magic mechanism relies on `_props.py`, which you'll need to update for the mesh visual. Once that's done, you should be able to just write `self.shine = value`.

It's definitely not ideal, but we'll revisit this later.

ron-adam commented:
Yes, I was thinking about the magic the same way you are.   But I'm not
quite far enough along in knowing datoviz to suggest an alternative yet.

I should be able to finish this up.  Or at least get to the point where you
only need a few edits.

I'll look at _prop.py.  I think that was the part I was missing.

On Mon, May 19, 2025, 1:51 PM Cyrille Rossant ***@***.***>
wrote:

> *rossant* left a comment (datoviz/datoviz#86)
> <https://github.com/datoviz/datoviz/pull/86#issuecomment-2891982079>
>
> I'm not really happy with the current implementation of visuals.py —
> there's too much hidden magic. Most setter functions just do self.xxx =
> yyy, which is intercepted by __setattr__() and automatically calls
> something like dvz.thevisualname_xxx(..., yyy). See the Prop
> implementation for reference.
>
> The issue is that this doesn't work in all cases, since some functions
> require more complex logic. But in your case, it ought to work.
>
> This magic mechanism relies on _props.py, which you'll need to update for
> the mesh visual. Once that's done, you should be able to just write self.shine
> = value.
>
> It's definitely not ideal, but we'll revisit this later.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/datoviz/datoviz/pull/86#issuecomment-2891982079>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AW7MURVUBSOFO6STCHN4QQT27IR4DAVCNFSM6AAAAAB5FCPX62VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQOJRHE4DEMBXHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ron-adam commented:
Update... Have it working, but need to fix some examples and affected files.

There will be some things still to fix and/or change.   And possibly some design decisions to make.

I'm not happy with how light colors and positions are set currently.   I'm thinking it needs to be both methods (set_dir, and set_pos)  using vec3 types,  but they will then set the w to 0 or 1 component to indicate to the shader that it's either a position or a direction.  If it's a direction (w=0), it get's normalized in the shader.  And the lights are similar, if the alpha for the light is 0, it is off.   An off light does a `continue` in the light loop, so it skips the lighting calculations for it.   But in both of these cases, it may be better to separate it out as a flag or switch that gets passed to the C set function.  

I still need to update the tests, and check check for documentation changes.   But there may be things you want to do differently so I'll wait on your feed back for those.   Apologies for taking a bit longer on this.   However it definitely helped me understand quite a bit of how the code works and is arranged.   (And was a good (needed) refresher for my C programing skills.)

I'll do a push tomorrow and you can look it over.

ron-adam commented:
As I said there is still some more to do.  Hopefully I got most of it correct.

Not sure about the conflicts.  Maybe you can look at those. 

Instead of naming it advanced light in the structures, I just used "light".  My thinking is this should work as the default light,  and the smaller possibly faster version can be an option set with a flag on the panel. 

Let me know if I need to change anything.  Meanwhile, I'm going over the files and fixing the parts I missed and making some small adjustments where it's needed.

ron-adam commented:
Do you have a preference for what default light position(s) and colors should be?   It's currently not setting the defaults correctly, but it shouldn't be too difficult to fix.

Questions... where should the light default values be?   Keeping them local is easier to manage for now, but global can work too.

Once the defaults are set properly the examples can use it instead of setting all the values.   That will make the examples more like what they were earlier.

rossant commented:
Can the default light positions/directions/parameters be set here? https://github.com/datoviz/datoviz/blob/dev/include/datoviz_enums.h#L929

ron-adam commented:
Ok,  that will work.

ron-adam commented:
I tend to have fresh ideas after waking up.

This morning's idea  (for a later date/PR) was to rework the light and mesh_light examples so that the light.py only does light configurations, with an option to load and save the configuration to a resource directory.   And to change mesh_light to work with only material configurations  (rename to material.py) , also being able to load and save them.  These both may get more complex as we add features, so keeping them separated makes each panel more focused.   Possibly add an object example selector so you can see how it looks on different kinds of objects.

(New thought)... these can be made into prebuilt panels that can be added to an application like the color picker, but without the save and load capability.  The file operations should be external to the panels.

ron-adam commented:
Thanks I'll look at the wolfram lights tonight.

I'm currently switching the sphere visuals to use the advance lights.  (And the basics lights will be unused, but still present.)
I should be finished with it tonight so it will be ready for you in morning. (Likely when you read this.)   This means both meshs and spheres will use the same lighting and include textures.. (Although textures may not work initially)   I need to adjust the light shader a bit.  The Wolfram link may help with that.

Here you can see the lights reflected in the spheres.   Tests with my sphere animation program didn't see any significant speed reductions,  So I think this will work out well.   (We probably just need to be generally correct.  I'm sure we will need to make small adjustments as we continue to work on it.)

![image](https://github.com/user-attachments/assets/7f12058c-5dd0-44a3-a57b-29111091428d)

ron-adam commented:
Ok,  you can take over.   (I'll be asleep for the next 7 or 8 hours.)

There is an emit value that does not update, not sure why.  It was working earlier.  See the mesh_light.py example.   I realized today that 'emit' posable wasn't a good choice due to you having used that term for other purpose. 

The emission feature allows an object to light up as if it's emitting it's own light.  The color is independent of it's color and other lights.  It may not be very useful as it changes all the objects at once.  It was an alternative to using high values of ambient for emission effect.  We can remove it.

I left the defaults colors and positions I was using to test it with.  (all the lights are on.)  They can be switched to off by setting the default light color alpha value to 0.   Also these values were just used to test it.  No need to keep them if you want something else.

rossant commented:
Excellent, thanks for your hard work on this! I'm going to take over now.

rossant commented:
Status update: I resolved the conflicts, rebased on dev, and I'm making good progress integrating this PR. It's not fully finished yet. I found a few existing bugs that were brought to light (pun intended) by this PR. If you're okay with it, I can push my changes to your branch, but it will be a destructive force push. It's better if you avoid committing to this branch until I'm done. You may need to delete your local branch and check it out again after the force push. Alternatively, I can close this PR and open a new one based on the rebased branch.

ron-adam commented:
Go ahead and do what ever you think is best.

I am expecting you to change some of it.  It will be interesting to me to
see how and what you do with this PR.  :-)

On Fri, May 23, 2025, 8:32 AM Cyrille Rossant ***@***.***>
wrote:

> *rossant* left a comment (datoviz/datoviz#86)
> <https://github.com/datoviz/datoviz/pull/86#issuecomment-2904437858>
>
> Status update: I resolved the conflicts, rebased on dev, and I'm making
> good progress integrating this PR. It's not fully finished yet. I found a
> few existing bugs that were brought to light (pun intended) by this PR. If
> you're okay with it, I can push my changes to your branch, but it will be a
> destructive force push. It's better if you avoid committing to this branch
> until I'm done. You may need to delete your local branch and check it out
> again after the force push. Alternatively, I can close this PR and open a
> new one based on the rebased branch.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/datoviz/datoviz/pull/86#issuecomment-2904437858>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AW7MURV3RA2L3PUTPMPVIOD274PNRAVCNFSM6AAAAAB5FCPX62VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMBUGQZTOOBVHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

rossant commented:
Thank you. I'll try to find some time to continue tomorrow. There are some things that are more or less independent from this PR but still want to address right now as they'll simplify a few things in the PR.

ron-adam commented:
I suggest creating a light object interface. 

It may look like this.
```python
    lights = app.lights(...light setup here...)         # Not yet attached to a panel.

    # Access and modify lights attributes here if needed.

    panel.add(lights)         # Attach lights to a panel. (replaces the default (or other) light object reference.)

    # Also modify lights object here and have the changes visible in the panel.
```    
A lights object means we don't need to keep a separate data structure around. 

Then it is also possible to have several light objects and switch between them.   Or to load and save a preconfigured light objects.  It also keeps the light code in one place, so it will be easier to maintain.

Cameras and lights have some similarities.  So it's likely some of the camera code may be useful.

That is all...  You may already be thinking about this.

ron-adam commented:
I think you are close to 3.0, so it should wait.    

Yes, that API looks consistent with other parts.

rossant commented:
Okay, merged! A few notes on what I did:

* Fixed the conflicts
* Moved the `PROPS` constant from `_constants.py` to `_props.py`, which caused some of the conflicts
* Fixed the showcase examples using the mesh or sphere visuals
* Fixed the 2D mesh C tests (the issue was in the fragment shader). To run tests, you can use `DVZ_DEBUG=1 ./build/datoviz test test_mesh`. They're defined in `tests/scene/test_mesh.c`. We should probably add this to the contributors documentation
* The `DvzColor` logic (based on either `uint` or `float`) is a bit of a historical leftover, but I kept it for now. I tried to simplify the implementation. It's still not ideal, but good enough for now

You can check the `dev` branch now. There are still things to fix. I reverted to a single default light for now, but we can revisit that. Ideally, the lighting should be soft and non-distracting for scientific visualization, without being too dull.

We’re close to the release. Time for final testing, especially of the Python examples, last necessary fixes, and checking the (updated) documentation.

Next steps before starting the release process (see `MAINTAINERS.md`):

* Add a panel background in all examples
* Improve the gallery
* Create a new “hero” image for the README and main documentation page, based on gallery screenshots
* Review, proofread, and polish the user guide, adding images and screenshots

I'll be back in \~12 hours. In the meantime, feel free to test the Python examples and report any issues. Check the documentation too: you can read the markdown files in `docs/`, or run `just serve` to start a local version of the documentation website using `mkdocs-material`.

Feel free to open a new PR if needed.

Thanks again!
